token-swap/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+pub struct TokenSwap {
+    authority: Pubkey,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum State {
+    /// Unallocated state, may be initialized into another state.
+    Unallocated,
+    Init(TokenSwap),
+}
+
+struct Invariant {
+    token_a: u64,
+    token_b: u64,
+}
+
+impl Invariant {
+    fn swap(&mut self, token_a: u64) -> Option<u64> {
+        let invariant = self.token_a.checked_mul(self.token_b)?;
+        let new_a = self.token_a.checked_add(token_a)?;
+        let new_b = invariant.checked_div(new_a)?;
+        let remove = self.token_b.checked_sub(new_b)?;
+        self.token_a = new_a;
+        self.token_b = new_b;
+        Some(remove)
+    }
+}
+
+impl<'a> State {
+     pub fn to_str(key: &Pubkey) -> &str {
+        unimplemented!();
+     }
+     pub fn deserialize(input: &'a [u8]) -> Result<Self, ProgramError> {
+         if input.len() < size_of::<u8>() {
+             return Err(ProgramError::InvalidAccountData);
+         }
+         Ok(match input[0] {
+             0 => Self::Unallocated,
+             1 => {
+                 if input.len() < size_of::<u8>() + size_of::<TokenSwap>() {
+                     return Err(ProgramError::InvalidAccountData);
+                 }
+                 #[allow(clippy::cast_ptr_alignment)]
+                 let swap: &TokenSwap = unsafe { &*(&input[1] as *const u8 as *const TokenSwap) };
+                 Self::Init(*swap)
+             }
+             _ => return Err(ProgramError::InvalidAccountData),
+         })
+     }
+ 
+     pub fn serialize(self: &Self, output: &mut [u8]) -> ProgramResult {
+         if output.len() < size_of::<u8>() {
+             return Err(ProgramError::InvalidAccountData);
+         }
+         match self {
+             Self::Unallocated => output[0] = 0,
+             Self::Init(swap) => {
+                 if output.len() < size_of::<u8>() + size_of::<TokenSwap>() {
+                     return Err(ProgramError::InvalidAccountData);
+                 }
+                 output[0] = 1;
+                 #[allow(clippy::cast_ptr_alignment)]
+                 let value = unsafe { &mut *(&mut output[1] as *mut u8 as *mut TokenSwap) };
+                 *value = *token;
+             }
+         }
+         Ok(())
+     }
+ 
+     pub fn create_token_account(
+         kind: &str,
+         instance_id: &Pubkey,
+         program_id: &Pubkey,
+         token_account: &AccountInfo,
+     ) -> ProgramResult {
+         // create token A account
+         let instruction_data = vec![];
+         let instruction = token::instruction::Instruction::NewAccount;
+         instruction.serialize(&mut instruction_data)?;
+ 
+         let account_addr = Pubkey::create_program_address(&[Self::to_str(instance_id), kind], program_id)?;
+         let authority = Pubkey::create_program_address(&[Self::to_str(instance_id), "authority"], program_id)?;
+         let invoked_instruction = create_instruction(
```

Comment:
This is not a generalized helper to create instructions, as above you are referencing the `invoked` program's specific instruction creation function.  If you want to create an instruction you need to use the sdk's helpers.